examples/siamese_network/siamese_network.py

```diff
@@ -0,0 +1,284 @@
+        index_2 = self.grouped_examples[selected_class][random_index_2]
+
+        # get the second image
+        negative_image = self.data[index_2].clone().float()
+
+        return anchor_image, positive_image, negative_image, anchor_label
+
+
+def train(model, device, optimizer, train_loader, lr_scheduler, log_interval, max_epochs):
+
+    criterion = nn.TripletMarginLoss()
+
+    # define model training step
+    def train_step(engine, batch):
+        model.train()
+        anchor_image, positive_image, negative_image, anchor_label = batch
+        anchor_image = anchor_image.to(device)
+        positive_image, negative_image = positive_image.to(device), negative_image.to(device)
+        anchor_label = anchor_label.to(device)
+        optimizer.zero_grad()
+        anchor_out = model(anchor_image)
+        positive_out = model(positive_image)
+        negative_out = model(negative_image)
+        loss = criterion(anchor_out, positive_out, negative_out)
+        loss.backward()
+        optimizer.step()
+        return loss
+
+    # create a trainer engine and attach train_step
+    trainer = Engine(train_step)
+
+    # attach progress bar to trainer
+    pbar = ProgressBar()
+    pbar.attach(trainer)
+
+    # attach various handlers to trainer engine
+    @trainer.on(Events.ITERATION_COMPLETED(every=log_interval))
+    def log_training_results(engine):
+        print(f"Train Epoch: {engine.state.epoch}, Train Loss: {engine.state.output: .5f}")
+
+    trainer.add_event_handler(Events.ITERATION_STARTED, lr_scheduler)
+
+    # run trainer engine
+    trainer.run(train_loader, max_epochs=max_epochs)
+
+
+def test(model, device, test_loader, lr_scheduler, log_interval):
+
+    criterion = nn.TripletMarginLoss()
+    average_test_loss = 0
+
+    # define model testing step
+    def test_step(engine, batch):
+        model.eval()
+        anchor_image, positive_image, negative_image, anchor_label = batch
+        anchor_image = anchor_image.to(device)
+        positive_image, negative_image = positive_image.to(device), negative_image.to(device)
+        anchor_label = anchor_label.to(device)
+        anchor_out = model(anchor_image)
+        positive_out = model(positive_image)
+        negative_out = model(negative_image)
+        test_loss = criterion(anchor_out, positive_out, negative_out)
+        return test_loss
+
+    # create evaluator engine and attach test step
+    evaluator = Engine(test_step)
+
+    # attach progress bar to evaluator
+    pbar = ProgressBar()
+    pbar.attach(evaluator)
+
+    # attach various handlers to evaluator engine
+    @evaluator.on(Events.ITERATION_COMPLETED(every=log_interval))
+    def log_testing_results(engine):
+        nonlocal average_test_loss
+        average_test_loss += engine.state.output
+        print(f"Test Epoch: {engine.state.epoch} Test Loss: {engine.state.output: .5f}")
+
+    evaluator.add_event_handler(Events.ITERATION_STARTED, lr_scheduler)
```

Comment:
Why do you update lr_scheduler during the testing ?

Reply:
I overlooked that detail, will remove it now.